sympy/functions/combinatorial/factorials.py

```diff
@@ -846,18 +846,24 @@ def _eval(self, n, k):
     def eval(cls, n, k):
         n, k = map(sympify, (n, k))
         d = n - k
+        is_int = (n.is_integer and k.is_integer)
         if d.is_zero or k.is_zero:
             return S.One
         elif d.is_zero is False:
             if (k - 1).is_zero:
                 return n
             elif k.is_negative:
-                return S.Zero
-            elif n.is_integer and n.is_nonnegative and d.is_negative:
+                if (is_int == S.true or is_int is None) and (n.is_nonnegative or n < k):
+                    return S.Zero
+                if is_int == S.false and k.is_integer:
+                    return S.Zero
+            elif n.is_integer and n.is_nonnegative and d.is_negative and \
+                (is_int == S.true or is_int is None):
```

Comment:
This is complicated. Can you explain in the text of pull request why this is the right combination of conditions to use? What is the source of this definition of binomial?

Reply:
SymPy `is_~` methods return Python `bool`.

Reply:
> This is complicated

if n is positive and d is negative then it means k>n(k is positive) hence return 0, this only implies for integers(therefore the check `is_int` is used)